lib/coverage/enabled.ml

```diff
@@ -0,0 +1,117 @@
+      tmp
+    with e ->
+      Sys.remove tmp;
+      D.warn "Failed to save coverage: %s" (Printexc.to_string e);
+      raise e
+
+  let reset () =
+    Bisect.Runtime.reset_counters ();
+    D.debug "Coverage counters reset"
+
+  let init_env name =
+    let (//)    = Filename.concat in
+    let tmpdir  = Filename.get_temp_dir_name () in
+    try
+      ignore (Sys.getenv bisect_file)
+    with Not_found ->
+      Unix.putenv bisect_file (tmpdir // Printf.sprintf "bisect-%s-" name)
+
+  let process body =
+    match Stringext.split ~on:' ' body with
+    | ["reset"] -> reset (); ""
+    | ["dump"; jobid] -> jobid |> Int64.of_string |> dump
+    | _ -> failwith body
+
+  let init name =
+    init_env name;
+    let queue_name = prefix ^ "." ^ name in
+    let (_:Thread.t) =
+      Thread.create (Protocol_unix.Server.listen ~process
+                       ~switch:!Xcp_client.switch_path ~queue:queue_name) () in
+    D.debug "Started coverage API thread on %s" queue_name;
+    ()
+end
+
+module Dispatcher = struct
+  let self = prefix ^ ".dispatch"
+  open Protocol_unix
+
+  let rpc_ignore_err ~t ~body queue =
+    D.debug "Dispatching %s to %s" body queue;
+    match Client.(rpc ~t ~queue ~body () |> error_to_msg) with
+    | `Ok x -> x
+    | `Error (`Msg e) ->
+       D.info "Failed to get coverage data from %s: %s" queue e;
+       ""
+
+  let string_of_result = function
+    | `Ok s -> s
+    | `Error (`Msg e) ->
+       D.info "Failed to get coverage data: %s" e;
+       "ERROR"
+
+  let process = fun body ->
+    let open Message_switch.Mresult in
+    D.debug "Coverage dispatcher received %s" body;
+    let result = begin
+        Client.connect ~switch:!Xcp_client.switch_path () >>= fun t ->
+        Client.list ~t ~prefix ~filter:`Alive () >>= fun queues ->
+        queues |>
+          (* filter out ourselves *)
+          List.filter (fun q -> self <> q) |>
+
+          (* best-effort: collect and return all non-failed results, log errors *)
+          List.rev_map (rpc_ignore_err ~t ~body) |>
+
+          (* multiple return values converted to a single string, suitable for use in a command like:
+          mv $(message-cli call org.xen.xapi.coverage.dispatch --timeout 60 --body 'dump {jobid}') /tmp/coverage/
+           *)
+          String.concat " " |>
+          ok
+      end |> Client.error_to_msg |> string_of_result in
+    D.debug "Coverage dispatcher replying to '%s': %s" body result;
+    result
+
+  let init () =
+    (* receives command and dispatches to all other coverage message queues *)
+    let (_:Thread.t) =
+      Thread.create (Protocol_unix.Server.listen ~process
+                       ~switch:!Xcp_client.switch_path ~queue:self) () in
+    D.debug "Started coverage API dispatcher on %s" self;
+    ()
+end
+
+(** [init name] sets up coverage profiling for binary [name]. You could 
+ *  use [Sys.argv.(0)] for [name].
+ *)
+let init name =
+  D.info "Coverage runtime initialized";
```

Comment:
Would it be better to swap these two lines so that we log the message only if (and only after) the initialisation succeeds?
(Or change the message to "About to..." or "Initialising coverage runtime.")

Reply:
I've added 'about to...', more logging is always good, and I want to be aware that any of the coverage code is enabled.